driver/pelican/types/scheduleSet.go

```diff
@@ -0,0 +1,88 @@
+package types
+
+import (
+	"encoding/xml"
+	"fmt"
+
+	"github.com/parnurzeal/gorequest"
+	rrule "github.com/teambition/rrule-go"
+)
+
+func (pel *Pelican) SetSchedule(newSchedule *ThermostatSchedule) error {
+	// Retrieve thermostat's latest schedule for extra block deletion purposes
+	currentSchedule, currentErr := pel.GetSchedule()
+	if currentErr != nil {
+		return fmt.Errorf("Error retrieving thermostat %v's current schedule: %v", pel.Name, currentErr)
+	}
+
+	for day, blocks := range newSchedule.DaySchedules {
+		// Delete Unnecessary Blocks + Error Checking
+		currentBlockCount := len(currentSchedule.DaySchedules[day])
+		requiredBlockCount := len(blocks)
+		for requiredBlockCount < currentBlockCount {
+			respDelete, _, errsDelete := gorequest.New().Get(pel.target).
+				Param("username", pel.username).
+				Param("password", pel.password).
+				Param("request", "set").
+				Param("object", "thermostatSchedule").
+				Param("selection", fmt.Sprintf("name:%s;dayOfWeek:%s;setTime:%v;", pel.Name, day, currentBlockCount)).
+				Param("value", "delete").
+				End()
+
+			if errsDelete != nil {
+				return fmt.Errorf("Error deleting thermostat schedule settings on day %v: %v\n", day, errsDelete)
+			}
+			var resultDelete apiResult
+			decDelete := xml.NewDecoder(respDelete.Body)
+			if errDecodeDelete := decDelete.Decode(&resultDelete); errDecodeDelete != nil {
+				return fmt.Errorf("Failed to decode thermostat schedule delete response XML: %v", errDecodeDelete)
+			}
+			if resultDelete.Success == 0 {
+				return fmt.Errorf("Error deleting thermostat schedule settings on day %v: %v\n", day, resultDelete.Message)
+			}
+			defer respDelete.Body.Close()
```

Comment:
Let's move this `defer` to before we do any decoding, i.e. after checking that `errsDelete` is `nil`.

Reply:
Got it, makes sense! The idea is that I'm deferring the close to the end, so I should probably place it earlier in the code haha, thanks for the catch!